### ISSUE TITLE
Implement EIP 1024 encryption methods

### ISSUE THREAD

**Issue by topealabi** [openethereum/openethereum#9893]
- **Parity Ethereum version**: 2.1.4
- **Operating system**: all
- **Installation**: all
- **Fully synchronized**: yes
- **Network**: all
- **Restarted**: yes

## Current Behavior

Geth does not currently provide any encryption methods

Parity currently provides a data encryption method using elliptic curve secp256k1. 

## Wanted Behavior

A new encryption method that works across different key management softwares. For example a sender could encrypt data with their MetaMask wallet and the receiver could decrypt it with Parity. This new encryption method will use curve25519 as specified in [EIP 1024](https://github.com/ethereum/EIPs/pull/1098). 

 [EIP 1024](https://github.com/ethereum/EIPs/pull/1098) uses a well audited library called [nacl](https://nacl.cr.yp.to/). The rust implementation of the necessary nacl functions can be found [here](https://github.com/javadevelopr/nacl-mini) 

##Definition of Done

You have successfully completed this bounty, when:

1. A developer can request an accounts encryption public key
2. A developer can encrypt any arbitrary data with using elliptical curve25519
3. A developer can decrypt any arbitrary data encrypted with their accounts encryption public key
4. These methods are compatible with MetaMasks encryption methods listed (here)[https://github.com/MetaMask/eth-sig-util]

**Comment by tomusdrw** [comment]
What's the motivation behind using a different curve? I think currently it's really nice that security-wise, ethereum pretty much relies only on keccak & secp256k1. Using another curve requires additional (potentially unsafe) crypto libraries (or using bindings, which we would really want to avoid) and complicates the case with public keys (a need to communicate them somehow).

**Comment by topealabi** [comment]
Hi @tomusdrw this suggestion was made by [crypto-experts](https://github.com/MetaMask/eth-sig-util/pull/18#issuecomment-384796354) (I am not a crypto-expert) who suggest that using the same curve for signing and encryption increases the surface area of attack. The [nacl](https://nacl.cr.yp.to/) library, which is the library suggested in the EIP, is one of the most well audited crypto libraries available. While there is no official rust version of nacl, I know of one developer who has already begun writing the necessary [nacl functions in rust](https://github.com/javadevelopr/nacl-mini).

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 1000.0 DAI (1000.0 USD @ $1.0/DAI)  attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $38,265.06 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by DanielShteinbok** [comment]
For number 3, did you mean to say private key?

**Comment by DanielShteinbok** [comment]
Pardon me, but did you mean to say "private key" for number 3?

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__Workers have applied to start work__.


These users each claimed they can complete the work by 2 months, 4 weeks from now.
Please review their action plans below:


**1) [shamardy](https://gitcoin.co/profile/shamardy) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?mutate_worker_action=approve&worker=shamardy) | [reject worker](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?mutate_worker_action=reject&worker=shamardy))_.**

I would like to work on this issue.

In addition to the 3 required methods for requesting an accounts encryption public key, encrypting and decrypting data. I think there should be a method for the generation of curve25519 keypairs from the Ethereum private key.

Also which namespace do we use for these methods, do we use the eth_ namespace?

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817).

**Comment by tomusdrw** [comment]
@topealabi thanks for the details, I wonder however what's the reasoning behind having that implemented in the node itself - it's not part of the protocol, it's strictly something that is used on top.
We've already removed high-level languages compilation from the RPCs, we are planning on removing accounts altogether, I don't really know if such functionality fits (already bloated) nodes APIs.

For the record, I'd be really hesitant of using `nacl-mini`, the gh profile of the owner looks fishy.

@shamardy:
1. Indeed generating keypairs would be nice addition, perhaps in `parity_accouns` namespace or `personal_` (under a flag)
1. `eth_` namespace is fine, but please hide it under `--experimental-rpcs` (perhaps we should extend that in future to be more granular). Would be good to have the proposed methods names in the EIP as well.

**Comment by topealabi** [comment]
Haha @tomusdrw I know the developer of `nacl-mini` in real life, he’s not shady he’s just shy and thinks GitHub is like a social network or something.

If you are still weary we can use a library like ring or rust-crypto for the encryption as long as we can decrypt using another key manger like Metamask. The whole point of this is to be cross-client. Other than that we are happy to take your guidance on where this should live in the Parity stack.

**Comment by ceresstation** [comment]
Hey @topealabi do you have a preference for who works on this? It looks like @tomusdrw commented pretty quickly but @shamardy was the first to formally apply.

**Comment by Tbaut** [comment]
Hey @ceresstation, you surely meant @zachzundel as @tomusdrw works at Parity and is overlooking this issue.

Please make sure we are carefully aligned on the task/tools before starting any work on this.

**Comment by ceresstation** [comment]
Yep, thanks for catching that @tbaut, just updated, @topealabi @tomusdrw I'll leave it to you both to confirm a candidate :)

**Comment by ceresstation** [comment]
Hey @zachzundel looks like @topealabi accepted you, you're good to go :)

**Comment by topealabi** [comment]
@zachzundel I think we should use ring, also the Parity folks have promised to provide guidance with this so perhaps we can create a slack channel on gitcoin for higher throughput communication

**Comment by Tbaut** [comment]
@tomusdrw is currently OOO, hence the non-responsiveness, please wait for his guidance before starting anything.

**Comment by kdenhartog** [comment]
> Hi @tomusdrw this suggestion was made by [crypto-experts](https://github.com/MetaMask/eth-sig-util/pull/18#issuecomment-384796354) (I am not a crypto-expert) who suggest that using the same curve for signing and encryption increases the surface area of attack. The [nacl](https://nacl.cr.yp.to/) library, which is the library suggested in the EIP, is one of the most well audited crypto libraries available. While there is no official rust version of nacl, I know of one developer who has already begun writing the necessary [nacl functions in rust](https://github.com/javadevelopr/nacl-mini).

We've been building this functionality in rust based on sodium-oxide which wraps libsodium through FFI calls. Are you looking for a pure rust implementation of nacl?

**Comment by gitcoinbot** [comment]
@zachzundel Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=1) | [3 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=3) | [5 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=5) | [10 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=10) | [100 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=100)

**Comment by zachzundel** [comment]
Yes, sorry, I am still working on this. I will plan on having a WIP PR out this weekend.

**Comment by gitcoinbot** [comment]
@zachzundel Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=1) | [3 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=3) | [5 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=5) | [10 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=10) | [100 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=100)

**Comment by gitcoinbot** [comment]
@zachzundel Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=1) | [3 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=3) | [5 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=5) | [10 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=10) | [100 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=100)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

@zachzundel due to inactivity, we have escalated [this issue](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817) to Gitcoin's moderation team. Let us know if you believe this has been done in error!

* [x] reminder (3 days)
* [x] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=1) | [3 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=3) | [5 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=5) | [10 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=10) | [100 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=100)

**Comment by zachzundel** [comment]
Sorry! I am still working on this!! If I haven’t made any progress in a
couple days I will release it back to the Gitcoin pool

**Comment by topealabi** [comment]
@zachzundel  awesome, let us know how we can help

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

@zachzundel due to inactivity, we have escalated [this issue](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817) to Gitcoin's moderation team. Let us know if you believe this has been done in error!

* [x] reminder (3 days)
* [x] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=1) | [3 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=3) | [5 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=5) | [10 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=10) | [100 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=100)

**Comment by zachzundel** [comment]
Pushed a WIP PR. I am still getting up to speed on Rust and the Parity codebase, so I probably made some mistakes and would appreciate a look over.

**Comment by gitcoinbot** [comment]
@zachzundel Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=1) | [3 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=3) | [5 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=5) | [10 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=10) | [100 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=100)

**Comment by gitcoinbot** [comment]
@zachzundel Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=1) | [3 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=3) | [5 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=5) | [10 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=10) | [100 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=100)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

@zachzundel due to inactivity, we have escalated [this issue](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817) to Gitcoin's moderation team. Let us know if you believe this has been done in error!

* [x] reminder (3 days)
* [x] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=1) | [3 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=3) | [5 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=5) | [10 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=10) | [100 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=100)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

@zachzundel due to inactivity, we have escalated [this issue](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817) to Gitcoin's moderation team. Let us know if you believe this has been done in error!

* [x] reminder (3 days)
* [x] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=1) | [3 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=3) | [5 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=5) | [10 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=10) | [100 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=100)

**Comment by zachzundel** [comment]
I am going to relinquish this issue back into the Gitcoin pool. I don't have the time to sort it out.

**Comment by shamardy** [comment]
I would like to continue the work on this issue. I already applied to work on this from gitcoin before.

**Comment by topealabi** [comment]
Thanks for your efforts @zachzundel 

@shamardy I will work with the Gitcoin team to get you on-boarded

**Comment by gitcoinbot** [comment]
@shamardy Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=1) | [3 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=3) | [5 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=5) | [10 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=10) | [100 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=100)

**Comment by ceresstation** [comment]
@shamardy @topealabi how are things going here?

**Comment by topealabi** [comment]
@shamardy

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

@shamardy due to inactivity, we have escalated [this issue](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817) to Gitcoin's moderation team. Let us know if you believe this has been done in error!

* [x] reminder (3 days)
* [x] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=1) | [3 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=3) | [5 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=5) | [10 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=10) | [100 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=100)

**Comment by shamardy** [comment]
@ceresstation @topealabi Still working on this issue. Should have a PR ready for review in a week.

**Comment by gitcoinbot** [comment]
@shamardy Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=1) | [3 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=3) | [5 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=5) | [10 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=10) | [100 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=100)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

@shamardy due to inactivity, we have escalated [this issue](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817) to Gitcoin's moderation team. Let us know if you believe this has been done in error!

* [x] reminder (3 days)
* [x] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=1) | [3 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=3) | [5 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=5) | [10 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=10) | [100 days](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817?snooze=100)

**Comment by hskang9** [comment]
Hello all, I am currently implementing EIP-1024 [here](https://github.com/hskang9/eip1024-rs/blob/master/src/lib.rs). So far the encryption and generating public key is implemented. Is this bounty still alive?

**Comment by hskang9** [comment]
Ok now done with test.

**Comment by topealabi** [comment]
Nice work @hskang9 , @tomusdrw can you take a look at this?

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__Work for 1000.0 DAI (1000.0 USD @ $1.0/DAI) has been submitted by__:

 1. [@hskang9](https://gitcoin.co/hskang9)



@ceresstation please take a look at the submitted work:
* [PR](https://github.com/paritytech/parity-ethereum/pull/11284) by @hskang9

<hr>

* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817)
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $100,850.56 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by dvdplm** [comment]
@topealabi @ceresstation I am bit hesitant to accept this EIP in the repo; there's the matter of post-bounty maintenance of the submitted code ofc. Furthermore, the EIP itself is not on the roadmap for upcoming HFs. 
It is obviously fine to submit a PR as part of the work to develop the EIP into something accepted by all client implementors, but before that happens I don't see this as something we will be able to merge.

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. Submitted 4. **Done** 

<hr>

__The funding of 1000.0 DAI (1000.0 USD @ $1.0/DAI)  attached to this issue has been approved & issued to @hskang9.__  

 
* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/paritytech/parity-ethereum/9893/1817)
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $73,929.35 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)
